geth-utils/gethutil/mpt/witness/prepare_witness.go

```diff
@@ -300,6 +337,170 @@ func prepareWitnessSpecial(testName string, trieModifications []TrieModification
 	StoreNodes(testName, nodes)
 }
 
+// For stack trie, we have the following combinations ([proofS] -> [proofC])
+//
+//	-[] [(empty)] -> [LEAF]
+//	-[] [LEAF] -> [EXT - BRANCH - LEAF]
+//	-[] [EXT - BRANCH] -> [EXT - BRANCH - LEAF]
+//	-[] [EXT - BRANCH] -> [BRANCH - LEAF] (modified extension)
+//	-[] [EXT] -> [BRANCH - BRANCH - BRANCH - LEAF] --> 144
+//	-[] [BRANCH - LEAF] -> [BRANCH - BRANCH - LEAF] (modified extension)
+//	-[] [BRANCH - BRANCH - (...BRANCH)] -> [BRANCH - BRANCH - (...BRANCH) - LEAF]
+//	-[] [BRANCH - BRANCH - LEAF ] -> [BRANCH - BRANCH - LEAF - BRANCH - LEAF]
+//	-[] [LEAF] -> [LEAF]
+//	-[] [EXT] -> [EXT]
+//	-[] [EXT - EXT] -> [EXT - EXT]
+//	-[] [EXT - LEAF] -> [LEAF]
+//	-[] [BRANCH - LEAF] -> [BRANCH - BRANCH - EXT - BRANCH - LEAF]
+//	-[] [BRANCH - BRANCH - EXT] -> [BRANCH - BRANCH - EXT - BRANCH - LEAF]
+//	-[] [BRANCH - BRANCH - EXT - BRANCH - (LEAF)] -> [BRANCH - BRANCH - EXT - BRANCH - EXT - BRANCH - LEAF]
+//	-[] [BRANCH - BRANCH - EXT - BRANCH - (...BRANCH)] -> [BRANCH - BRANCH - EXT - BRANCH - (...BRANCH) - LEAF]
+//	-[] [LEAF] -> [BRANCH - BRANCH - EXT - BRANCH - BRANCH - LEAF]
```

Comment:
yep! I copied it from `obtainTwoProofsAndConvertToWitness` to avoid creating bugs for existing state trie. Once everything is done, will do refactor and merge functions.

For the scenario, `[EXT] -> [BRANCH - BRANCH - BRANCH - LEAF]`, you can check `TestTransaction144` in my latest commit.

Reply:
@miha-stopar , I found the reason why.

At 144th tx, the branch is hashed in [L808](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/feat/%231752-fix-witness-generation-json-export/geth-utils/gethutil/mpt/trie/stacktrie.go#L808) and it won't go into [L842](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/feat/%231752-fix-witness-generation-json-export/geth-utils/gethutil/mpt/trie/stacktrie.go#L842). That's why we only got 1 single proof (the proof should be `hashedNode` instead of `ExtNode` as well).

I guess we can modify L842 and just skip the last node instead of whole proofs. Does it make sense to you? Another thing is I noticed the last key is removed in [`TryUpdate`](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/feat/%231752-fix-witness-generation-json-export/geth-utils/gethutil/mpt/trie/stacktrie.go#L200). So, I was also thinking, maybe we could remove last key (which is terminal char, `16`) in the proofS ([here](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/feat/%231752-fix-witness-generation-json-export/geth-utils/gethutil/mpt/trie/stacktrie.go#L692)) before we update it. Wondering which method is more feasibe for you?

Reply:
Update:

Removed `isHashed` check in `GetProof` which makes normal proofs hashed node proofs concatenated together and the proofs looks more sense now. Have no more `[EXT] -> [BRANCH - BRANCH - BRANCH - LEAF]` or  `[LEAF] -> [BRANCH - BRANCH - EXT - BRANCH - BRANCH - LEAF]` these kinds of proofs.

Reply:
Yeah, the problem is part of the proof can be hashed and part of the proof not hashed.

I think the simplest way is to have `isHashed` flag for each node (having array `isHashed` which is of the same length as `nodes`). For those nodes that have already been hashed in the first loop, we prepare the RLPs (by RLPs I mean raw data) in the first loop and store them in the array, for example `endProof`.

For the nodes that haven't been hashed in the first loop (for these we have `isHashed = false`), we retrieve their RLPs in the second loop, we store them in something like `startProof`. Finally we concatenate `startProof` and `endProof`.

So not much changes, just assigning `isHashed` for each level in the first loop and checking `isHashed` for each level in the second loop (and either reusing the RLP provided in the first loop if `isHashed=true` or retrieving via `db.Get` if `isHashed=false`).

Reply:
Hey @miha-stopar 

As our last disccusion, you suggested not to handle modification extension node first and take care normal cases first. Would like to confirm my understanding with you and if so, I think this PR is ready for review.

Regarding to this `isHashed` issue, I pass the proof type directly. I think it's easier to tell the type when generating witness. Wondering if it makes sense to you or not.

Reply:
Got it. I'll run all the MPT tests again.

To be honest, I don't have any plans to merge `GenerateWitness` and `obtainTwoProofsAndConvertToWitness`. The code structure is simliar but the details inside is quite different. For example, there are some special cases for the first 2 txs and how we set `additionalBranch` is different. I'm afraid that if we merge these two functions into one, there will be many `if-else` and flags to distinguish state trie and stack trie. I don't have confidence to handle it right now but yes, we could leave it later to see if we could figure out a better way to refactor these two functions.

I have manually merged some of changes from #1756 related to [this discussion](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1756/files#r1565140245).  The scope of #1756 looks quite big, including 30 files, testing about wrong ext node and some circuit logic. Especially, it touches circuit files so I prefer not to rebase on #1756. But if any code is related to witness generation and also helpful to this PR, I would love to cherry-pick those commits.

Reply:
Agree, let's don't rebase it. For `GenerateWitness` and `obtainTwoProofsAndConvertToWitness` - I feel the implementation could stay pretty much the same, but I might be wrong, let's leave this for future PRs.

I will approve it, but will wait for the MPT tests to be checked to be on the safe side.